ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditTextInputConnectionWrapper.java

```diff
@@ -0,0 +1,164 @@
+ * TextInput has the property 'autoCorrect' set to true. When text is being composed in the composing
+ * state within the EditText, each key press will result in a call to
+ * {@link InputConnection#setComposingText(CharSequence, int)} with a CharSequence argument equal to
+ * that of the entire composing region, rather than a single character diff.
+ * We can reason about the keyPress based on the resultant cursor position changes of the EditText after
+ * applying this change. For example if the cursor moved backwards by one character when composing,
+ * it's likely it was a delete; if it moves forward by a character, likely to be a key press of that character.
+ *
+ * IMEs can also call {@link InputConnection#beginBatchEdit()} to signify a batch of operations. One
+ * such example is committing a word currently in composing state with the press of the space key.
+ * It is IME dependent but the stock Android keyboard behavior seems to be to commit the currently composing
+ * text with {@link InputConnection#setComposingText(CharSequence, int)} and commits a space character
+ * with a separate call to {@link InputConnection#setComposingText(CharSequence, int)}.
+ * Here we chose to emit the last input of a batch edit as that tends to be the user input, but
+ * it's completely arbitrary.
+ *
+ * Another function of this class is to detect backspaces when the cursor at the beginning of the
+ * {@link android.widget.EditText}, i.e no text is deleted.
+ *
+ * N.B. this class is only applicable for soft keyboards behavior. For hardware keyboards
+ * {@link android.view.View#onKeyDown(int, KeyEvent)} can be overridden to obtain the keycode of the
+ * key pressed.
+ */
+class ReactEditTextInputConnectionWrapper extends InputConnectionWrapper {
+  public static final String NEWLINE_RAW_VALUE = "\n";
+  public static final String BACKSPACE_KEY_VALUE = "Backspace";
+  public static final String ENTER_KEY_VALUE = "Enter";
+
+  private ReactEditText mEditText;
+  private EventDispatcher mEventDispatcher;
+  private boolean mIsBatchEdit;
+  private @Nullable String mKey = null;
+
+  public ReactEditTextInputConnectionWrapper(
+      InputConnection target,
+      boolean mutable,
+      final ReactContext reactContext,
+      final ReactEditText editText
+  ) {
+    super(target, mutable);
+    mEventDispatcher = reactContext.getNativeModule(UIManagerModule.class).getEventDispatcher();
+    mEditText = editText;
+  }
+
+  @Override
+  public boolean beginBatchEdit() {
+    mIsBatchEdit = true;
+    return super.beginBatchEdit();
+  }
+
+  @Override
+  public boolean endBatchEdit() {
+    mIsBatchEdit = false;
+    if(mKey != null) {
+      dispatchKeyEvent(mKey);
+      mKey = null;
+    }
+    return super.endBatchEdit();
+  }
+
+  @Override
+  public boolean setComposingText(CharSequence text, int newCursorPosition) {
+    final int previousSelectionStart = mEditText.getSelectionStart();
```

Comment:
can you drop all the `final` keywords here. It is not common to use `final` in this context in Java (unlike in JS where you'd use `const`). Usually `final` is only used for scoped vars when they are being referenced in anonymous classes ("lambda classes") – in which case you get a compiler error if you forget to use `final`

Reply:
Thank you!